roofit/histfactory/src/HistoToWorkspaceFactoryFast.cxx

```diff
@@ -737,82 +709,73 @@ namespace HistFactory{
+      firstHistFunc = dynamic_cast<const RooHistFunc*>(piecewiseInt->nominalHist());
     }
+    assert(firstHistFunc);
 
-    //vector<string>::iterator it=syst_x_expectedPrefixNames.begin();
-    for(unsigned int j=0; j<syst_x_expectedPrefixNames.size();++j){
-      std::stringstream str;
-      str<<"_"<<j;
-      // repatative, but we need one coeff for each term in the sum
-      // maybe can be avoided if we don't use bin width as coefficient
-      command=string(Form("binWidth_%s_%d[%e]",obsNameVecStr.c_str(),j,binWidth));     
-      proto->factory(command.c_str());
-      proto->var(Form("binWidth_%s_%d",obsNameVecStr.c_str(),j))->setConstant();
-      coeffList+=prepend+"binWidth_"+obsNameVecStr+str.str();
-
-      command="prod::L_x_"+syst_x_expectedPrefixNames.at(j)+"("+normByNames.at(j)+","+syst_x_expectedPrefixNames.at(j)+")";
-      /*RooAbsReal* tempFunc =(RooAbsReal*) */
-      proto->factory(command.c_str());
-      shapeList+=prepend+"L_x_"+syst_x_expectedPrefixNames.at(j);
-      prepend=",";
-      
-      // add to num int to product
-      //      tempFunc->specialIntegratorConfig(kTRUE)->method1D().setLabel("RooBinIntegrator")  ;
-      //      tempFunc->forceNumInt();
+    // Prepare a function to divide all bin contents by bin width to get a density:
+    const std::string binWidthFunctionName = totName + "_binWidth";
+    RooBinWidthFunction binWidth(binWidthFunctionName.c_str(), "Divide by bin width to obtain probability density", *firstHistFunc, true);
+    proto->import(binWidth);
+    auto binWidthWS = proto->function(binWidthFunctionName.c_str());
+    assert(binWidthWS);
 
-    }    
+    // Loop through samples and create products of their functions:
+    RooArgSet coefList;
+    RooArgSet shapeList;
+    for (unsigned int i=0; i < sampleHistFuncs.size(); ++i) {
+      assert(!sampleHistFuncs[i].empty());
+      coefList.add(*sampleScaleFactors[i]);
 
-    proto->defineSet("coefList",coeffList.c_str());
-    proto->defineSet("shapeList",shapeList.c_str());
-    //    proto->factory(command.c_str());
-    RooRealSumPdf tot(totName.c_str(),totName.c_str(),*proto->set("shapeList"),*proto->set("coefList"),kTRUE);
+      std::vector<RooAbsArg*>& thisSampleHistFuncs = sampleHistFuncs[i];
+      thisSampleHistFuncs.push_back(binWidthWS);
+
+      if (thisSampleHistFuncs.size() == 1) {
+        // Just one function. Book it.
+        shapeList.add(*thisSampleHistFuncs.front());
+      } else {
+        // Have multiple functions. We need to multiply them.
+        std::string name = thisSampleHistFuncs.front()->GetName();
+        auto pos = name.find("Hist_alpha");
+        if (pos != std::string::npos) {
+          name = name.substr(0, pos) + "shapes";
+        } else if ( (pos = name.find("nominal")) != std::string::npos) {
+          name = name.substr(0, pos) + "shapes";
+        }
+
+        RooArgSet terms;
+        for (auto func : thisSampleHistFuncs) {
+          terms.add(*func);
+        }
+
```

Comment:
- Not c++11! **Don't do this in interfaces!!!** Not even when ROOT becomes c++14 ...
- Not sure if vectors automatically convert to spans?
- One could do this with `begin(), end()`, but I wouldn't add another constructor. You can do if you want, but I'll go now with `begin, end`.

Reply:
Right, it's not C++11. But could we not use the backport of `std::span` in ROOT, like you did for the RooSpan back in the day?
https://github.com/root-project/root/blob/master/roofit/batchcompute/inc/RooSpan.h#L20
But okay, maybe we should better not do interface changes to basic RooFit classes in this PR. If they aren't completely non-controversial. I'm taking my suggestion back.

Reply:
No, not in interfaces. Note that it's always `RooSpan` in all interfaces. Imagine this situation:
- We call something `std::xxx` in ROOT
- People include it, and it works, because that thing is not part of the `std::` (because it's c++11).
- People switch their project to `c++20`, because "Why not?". ROOT's interfaces are c++11, but that will be compatible.
- Now there's two definitions of `std::span` in the interfaces. Nothing will compile.

You simply *cannot* use those things in interfaces. If it's internal to ROOT, it's OK, because we can typedef or fix includes etc, but once something is in the interfaces, it's out in the world.
Our span thing works because once ROOT goes c++20, you just exchange the (non-standard) `std::span` inside RooSpan with the real thing, and fix a few lines of code in RooSpan if necessary. Interfaces remain unchanged, and everything will just work.

Reply:
I fixed the template `(begin, end)` constructor of `RooArgSet`, so I was now able to write almost the exact same line as you proposed.

Reply:
That's a nice solution! And thanks for explaining the situation with the `std::` backports in interfaces.